cabal-install/src/Distribution/Client/IndexUtils.hs

```diff
@@ -556,18 +561,54 @@ extractPkg verbosity entry blockNo = case Tar.entryContent entry of
 extractPrefs :: Tar.Entry -> Maybe [Dependency]
 extractPrefs entry = case Tar.entryContent entry of
   Tar.NormalFile content _
-     | FilePath.Posix.takeFileName entrypath == "preferred-versions"
+     | isPreferredVersions entrypath
     -> Just prefs
     where
       entrypath = Tar.entryPath entry
       prefs     = parsePreferredVersions content
   _ -> Nothing
 
+------------------------------------------------------------------------
+-- Filename and parsers for 'preferred-versions' file.
+--
+
+-- | Expected name of the 'preferred-versions' file.
+--
+-- Contains special constraints, such as a preferred version of a package
+-- or deprecations of certain package versions.
+--
+-- Expected format:
+--
+-- @
+-- binary > 0.9.0.0 || < 0.9.0.0
+-- text == 1.2.1.0
+-- @
+preferredVersions :: FilePath
+preferredVersions = "preferred-versions"
+
+-- | Does the given filename match with the expected name of 'preferred-versions'?
+isPreferredVersions :: FilePath -> Bool
+isPreferredVersions = (== preferredVersions) . takeFileName
+
+-- | Parse `preferred-versions` file, ignoring any parse failures.
+--
+-- To obtain parse errors, use 'parsePreferredVersionsWarnings'.
 parsePreferredVersions :: ByteString -> [Dependency]
-parsePreferredVersions = mapMaybe simpleParsec
-                       . filter (not . isPrefixOf "--")
-                       . lines
-                       . BS.Char8.unpack -- TODO: Are we sure no unicode?
+parsePreferredVersions = rights . parsePreferredVersionsWarnings
+
+-- | Parse `preferred-versions` file, collecting parse errors that can be shown
+-- in error messages. Error message is the line that failed to parse.
+parsePreferredVersionsWarnings :: ByteString -> [Either String Dependency]
+parsePreferredVersionsWarnings =
+  map parsePreference
+  . filter (not . isPrefixOf "--")
+  . lines
+  . BS.Char8.unpack -- TODO: Are we sure no unicode?
```

Comment:
This is actually wrong in my opinion. Can we fix that in this PR, or rather a follow-up?

Reply:
you may use `fromUTF8BS` here, I don't think we need to be more elaborate here.